wandb/sdk/wandb_run.py

```diff
@@ -2046,29 +2046,38 @@ def _on_ready(self) -> None:
         # object is about to be returned to the user, don't let them modify it
         self._freeze()
 
+    def _make_job_source_reqs(self) -> Tuple[List[str], Dict[str, Any], Dict[str, Any]]:
+        import pkg_resources
+
+        installed_packages_list = sorted(
+            f"{d.key}=={d.version}" for d in iter(pkg_resources.working_set)
+        )
+        input_types = TypeRegistry.type_of(self.config.as_dict()).to_json()
+        output_types = TypeRegistry.type_of(self.summary._as_dict()).to_json()
+
+        return installed_packages_list, input_types, output_types
+
     def _log_job(self) -> None:
         # don't produce a job if the run is sourced from a job
         if (
             self._launch_artifact_mapping.get(wandb.util.LAUNCH_JOB_ARTIFACT_SLOT_NAME)
             is not None
         ):
             return
-        artifact = None
-        input_types = TypeRegistry.type_of(self.config.as_dict()).to_json()
-        output_types = TypeRegistry.type_of(self.summary._as_dict()).to_json()
 
-        import pkg_resources
-
-        installed_packages_list = sorted(
-            f"{d.key}=={d.version}" for d in iter(pkg_resources.working_set)
-        )
+        artifact = None
+        (
+            installed_packages_list,
+            input_types,
+            output_types,
+        ) = self._make_job_source_reqs()
 
         for job_creation_function in [
             self._create_repo_job,
             self._create_artifact_job,
             self._create_image_job,
         ]:
-            artifact = job_creation_function(
+            artifact = job_creation_function(  # type: ignore
```

Comment:
why is this needed?

Reply:
flake 8: `error: Cannot call function of unknown type  [operator]`

Reply:
Makes sense!